pkg/authentication/scram/scram.go

```diff
@@ -47,9 +81,238 @@ func EnsureAgentSecret(getUpdateCreator secret.GetUpdateCreator, secretNsName ty
+	needToGenerateNewCredentials, err := needToGenerateNewCredentials(getUpdateCreator, user, mdb, password)
+	if err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// there are no changes required, we can re-use the same credentials.
+	if !needToGenerateNewCredentials {
+		zap.S().Debugf("Credentials have not changed, using credentials stored in: secret/%s", scramCredentialsSecretName(mdb.Name, user.Name))
+		return readExistingCredentials(getUpdateCreator, mdb.NamespacedName(), user.Name)
+	}
+
+	// the password has changed, or we are generating it for the first time
+	sha1Creds, sha256Creds, err := generateScramShaCredentials(user, password)
+	if err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	// create or update our credentials secret for this user
+	zap.S().Debugf("Generating new credentials and storing in secret/%s", scramCredentialsSecretName(mdb.Name, user.Name))
+	if err := createScramCredentialsSecret(getUpdateCreator, mdb.NamespacedName(), user.Name, sha1Creds, sha256Creds); err != nil {
+		return scramcredentials.ScramCreds{}, scramcredentials.ScramCreds{}, err
+	}
+
+	zap.S().Debugf("Successfully generated SCRAM credentials")
+	return sha1Creds, sha256Creds, nil
+}
+
+// needToGenerateNewCredentials determines if it is required to generate new credentials or not.
+// this will be the case if we are either changing password, or are generating credentials for the first time.
+func needToGenerateNewCredentials(secretGetter secret.Getter, user mdbv1.MongoDBUser, mdb mdbv1.MongoDB, password string) (bool, error) {
+	s, err := secretGetter.GetSecret(types.NamespacedName{Name: scramCredentialsSecretName(mdb.Name, user.Name), Namespace: mdb.Namespace})
+	if err != nil {
+		// haven't generated credentials yet, so we are changing password
+		if errors.IsNotFound(err) {
+			zap.S().Debugf("No existing credentials found, generating new credentials")
+			return true, nil
+		}
+		return false, err
+	}
+
+	existingSha1Salt := s.Data[sha1SaltKey]
+	existingSha256Salt := s.Data[sha256SaltKey]
+
+	// the salts are stored encoded, we need to decode them before we use them for
+	// salt generation
+	decodedSha1Salt, err := base64.StdEncoding.DecodeString(string(existingSha1Salt))
+	if err != nil {
+		return false, err
+	}
+	decodedSha256Salt, err := base64.StdEncoding.DecodeString(string(existingSha256Salt))
+	if err != nil {
+		return false, err
+	}
+
+	// regenerate credentials using the existing salts in order to see if the password has changed.
+	sha1Creds, sha256Creds, err := computeScramShaCredentials(user.Name, password, decodedSha1Salt, decodedSha256Salt)
+	if err != nil {
+		return false, err
+	}
+
+	existingSha1Creds, existingSha256Creds, err := readExistingCredentials(secretGetter, mdb.NamespacedName(), user.Name)
+	if err != nil {
+		return false, err
+	}
+
+	sha1CredsAreTheSame := reflect.DeepEqual(sha1Creds, existingSha1Creds)
```

Comment:
Is `DeepEqual` necessary here? `ScramCreds` have no nested types and can be compared using `==` if I'm not mistaken.

Reply:
you're dead right! 👍